### ISSUE TITLE
runtime: program exceeds 10000-thread limit

### ISSUE THREAD

**Issue by poiuty** [NebulousLabs/Sia#1775]
```
# siac version
Sia Client v1.2.1
Sia Daemon v1.2.1
```

```
runtime: program exceeds 10000-thread limit
fatal error: thread exhaustion

runtime stack:
runtime.throw(0xaa3788, 0x11)
	/usr/local/go/src/runtime/panic.go:596 +0x95
runtime.checkmcount()
	/usr/local/go/src/runtime/proc.go:519 +0xa4
runtime.mcommoninit(0xc457678800)
	/usr/local/go/src/runtime/proc.go:539 +0xa0
runtime.allocm(0xc42001a000, 0x0, 0xc400000000)
	/usr/local/go/src/runtime/proc.go:1341 +0x99
runtime.newm(0x0, 0xc42001a000)
	/usr/local/go/src/runtime/proc.go:1611 +0x39
runtime.startm(0xc42001a000, 0xe96700)
	/usr/local/go/src/runtime/proc.go:1698 +0x17e
runtime.handoffp(0xc42001a000)
	/usr/local/go/src/runtime/proc.go:1725 +0x55
runtime.retake(0x53f7ec8deb9d9, 0x0)
	/usr/local/go/src/runtime/proc.go:3909 +0x15d
runtime.sysmon()
	/usr/local/go/src/runtime/proc.go:3836 +0x1ea
runtime.mstart1()
	/usr/local/go/src/runtime/proc.go:1179 +0x11e
runtime.mstart()
	/usr/local/go/src/runtime/proc.go:1149 +0x64

goroutine 1 [chan receive, 8719 minutes]:
main.startDaemon(0x0, 0x0, 0xaa096a, 0xe, 0xa8fe07, 0x5, 0xa8fe0c, 0x5, 0x0, 0x7ffe3c9ebeea, ...)
	/home/luke/go/src/github.com/NebulousLabs/Sia/siad/daemon.go:293 +0xa23
main.startDaemonCmd(0xc420074fc0, 0xc4200b9880, 0x0, 0x2)
	/home/luke/go/src/github.com/NebulousLabs/Sia/siad/daemon.go:309 +0x6d
github.com/spf13/cobra.(*Command).execute(0xc420074fc0, 0xc420010100, 0x2, 0x2, 0xc420074fc0, 0xc420010100)
	/home/luke/go/src/github.com/spf13/cobra/command.go:648 +0x231
github.com/spf13/cobra.(*Command).ExecuteC(0xc420074fc0, 0xabc01a, 0x3b, 0xc420140960)
	/home/luke/go/src/github.com/spf13/cobra/command.go:734 +0x339
github.com/spf13/cobra.(*Command).Execute(0xc420074fc0, 0xe965a0, 0xaa57b3)
	/home/luke/go/src/github.com/spf13/cobra/command.go:693 +0x2b
main.main()
	/home/luke/go/src/github.com/NebulousLabs/Sia/siad/main.go:174 +0x7ac

goroutine 17 [syscall, 10709 minutes, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2197 +0x1

goroutine 5 [syscall, 10709 minutes]:
os/signal.signal_recv(0x0)
	/usr/local/go/src/runtime/sigqueue.go:116 +0x104
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x22
created by os/signal.init.1
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x41

goroutine 7 [IO wait, 1561 minutes]:
net.runtime_pollWait(0x7ff98abbf088, 0x72, 0x0)
	/usr/local/go/src/runtime/netpoll.go:164 +0x59
net.(*pollDesc).wait(0xc420142df8, 0x72, 0x0, 0xc45dbd8740)
	/usr/local/go/src/net/fd_poll_runtime.go:75 +0x38
net.(*pollDesc).waitRead(0xc420142df8, 0xffffffffffffffff, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:80 +0x34
net.(*netFD).accept(0xc420142d90, 0x0, 0xd870a0, 0xc45dbd8740)
	/usr/local/go/src/net/fd_unix.go:430 +0x1e5
net.(*TCPListener).accept(0xc42000e160, 0xc420032e70, 0x696dae, 0x455150)
	/usr/local/go/src/net/tcpsock_posix.go:136 +0x2e
net.(*TCPListener).Accept(0xc42000e160, 0xac7648, 0xc42e27cd20, 0xd8d7a0, 0xc42015c0c0)
	/usr/local/go/src/net/tcpsock.go:228 +0x49
net/http.(*Server).Serve(0xc420092b00, 0xd8cb60, 0xc42000e160, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:2643 +0x228
main.(*Server).Serve(0xc4200b9c40, 0x0, 0x0)
	/home/luke/go/src/github.com/NebulousLabs/Sia/siad/server.go:398 +0x44
main.startDaemon.func1(0xc4201483c0, 0xc4200b9c40)
	/home/luke/go/src/github.com/NebulousLabs/Sia/siad/daemon.go:129 +0x2b
created by main.startDaemon
	/home/luke/go/src/github.com/NebulousLabs/Sia/siad/daemon.go:130 +0x312

goroutine 8 [IO wait, 1 minutes]:
net.runtime_pollWait(0x7ff98abbefc8, 0x72, 0x0)
	/usr/local/go/src/runtime/netpoll.go:164 +0x59
net.(*pollDesc).wait(0xc420142ed8, 0x72, 0x0, 0xc42dbf6600)
	/usr/local/go/src/net/fd_poll_runtime.go:75 +0x38
net.(*pollDesc).waitRead(0xc420142ed8, 0xffffffffffffffff, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:80 +0x34
net.(*netFD).accept(0xc420142e70, 0x0, 0xd870a0, 0xc42dbf6600)
	/usr/local/go/src/net/fd_unix.go:430 +0x1e5
net.(*TCPListener).accept(0xc42000e190, 0x2, 0x2, 0xc4201b9eb8)
	/usr/local/go/src/net/tcpsock_posix.go:136 +0x2e
net.(*TCPListener).Accept(0xc42000e190, 0xc420148420, 0x0, 0xd90101, 0xc46b54e150)
	/usr/local/go/src/net/tcpsock.go:228 +0x49
github.com/NebulousLabs/Sia/modules/gateway.(*Gateway).permanentListen(0xc42006f1e0, 0xc4201484e0)
	/home/luke/go/src/github.com/NebulousLabs/Sia/modules/gateway/peers.go:90 +0x75
created by github.com/NebulousLabs/Sia/modules/gateway.New
	/home/luke/go/src/github.com/NebulousLabs/Sia/modules/gateway/gateway.go:313 +0x957
```

https://pastebin.com/raw/eh9ZUVAV

**Comment by S-anasol** [comment]
+same error after 1-2 day

v1.1.2

**Comment by DavidVorick** [comment]
I'm having the same issues on my machine, so at least I know where to look.

Pretty sure I know what the issue is, not sure how easy the fix is though.

**Comment by DavidVorick** [comment]
I believe that I've got code which fixes the problem. Hard to tell immediately b/c the threads take a few days to build up, but I've got a node running with the proposed fix and am monitoring it.

So far so good.

**Comment by S-anasol** [comment]
@DavidVorick maybe it will help.
I got this on node which used to get data for SiaHub, i.e. many API requests everyday.
Other nodes still didnt get this error.

So to make it faster crash you can try spam api with some requests.

**Comment by poiuty** [comment]
@DavidVorick, crash every day, please public possible fix. I can help test

**Comment by DavidVorick** [comment]
@poiuty the fix is a series of pull requests on our NebulousLabs/muxado package. Unfortunately testing is suggesting that I've solved a set of problems but not the ones we want solved.

@poiuty: are you running the CLI? If so, you can get a thread dump by running:

`siad > output.txt 2>&1`

That'll dump all output to a file. When Sia has been running for a while and is close to the 10000 thread limit, you can get a full stack dump by typing `ctrl+\`. That should have a bit of output for each thread, which means a rather large text file should be created. That text file will help a lot to debug.

I should also add that if you haven't upgraded to v1.2.1 you should do so. Though it's not perfect, it has a lot of fixes and improvements that aren't in earlier versions.

**Comment by chargr** [comment]
Running v1.2.2 - here is the full output.

[siad.log.gz](https://github.com/NebulousLabs/Sia/files/1029697/siad.log.gz)

**Comment by chargr** [comment]
@poiuty how many folders do you have?  since reducing 8 folders to 4 folders I haven't crashed (I was crashing every 2 - 3 days).

**Comment by poiuty** [comment]
@chargr 
```
siac host 
Host info:
	Connectability Status: Host appears to be working.

	Storage:      14.9996 TB (822.595 GB used)
	Price:        1000 SC / TB / Month
	Max Duration: 25 Weeks

	Accepting Contracts:  Yes
	Anticipated Revenue:  19.11 KS
	Locked Collateral:    3.14 KS
	Revenue:              1.05 KS

Storage Folders:
    Used          Capacity     % Used    Path
    275.360 GB    4.9999 TB    5.51      /home/sia/storage
    272.445 GB    4.9999 TB    5.45      /home/sia/storage2
    274.790 GB    4.9999 TB    5.50      /home/sia/storage3
```

After upgrade to 1.2.2 dont crash.

**Comment by chargr** [comment]
Still crashing for me on 1.2.2, although seemingly less frequent.

**Comment by poiuty** [comment]
Auto unlock wallet (linux).
Edit `/home/sia/unlock.txt` (set your pass password)
Create `/etc/cron.d/sia`
```
SHELL=/bin/sh
PATH=/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin
MAILTO=""
HOME=/

*/10 * * * * sia echo $(cat /home/sia/unlock.txt) | siac wallet unlock >/dev/null 2>&1
```

Restart cron `/etc/init.d/cron restart`

**Comment by poiuty** [comment]
Host http://sia.org.ru/host/1313, lose ~9KS

```
siac host -v
General Info:
	Connectability Status: Host appears to be working.

Host Internal Settings:
	acceptingcontracts:   Yes
	maxduration:          25 Weeks
	maxdownloadbatchsize: 17.83 MB
	maxrevisebatchsize:   17.83 MB
	netaddress:           188.244.40.69:9982 (manually specified)
	windowsize:           24 Hours

	collateral:       100 SC / TB / Month
	collateralbudget: 1 MS 
	maxcollateral:    5 KS Per Contract

	mincontractprice:          20 SC
	mindownloadbandwidthprice: 250 SC / TB
	minstorageprice:           100 SC / TB / Month
	minuploadbandwidthprice:   10 SC / TB

Host Financials:
	Contract Count:               10201
	Transaction Fee Compensation: 980 SC
	Potential Fee Compensation:   17.24 KS
	Transaction Fee Expenses:     0 H

	Storage Revenue:           289.8 SC
	Potential Storage Revenue: 1.893 KS

	Locked Collateral: 3.586 KS
	Risked Collateral: 253 SC
	Lost Collateral:   0 H

	Download Revenue:           592.6 mS
	Potential Download Revenue: 5.773 SC
	Upload Revenue :            3.022 SC
	Potential Upload Revenue:   7.196 SC

RPC Stats:
	Error Calls:        9
	Unrecognized Calls: 0
	Download Calls:     0
	Renew Calls:        2
	Revise Calls:       0
	Settings Calls:     479
	FormContract Calls: 9

Storage Folders:
    Used          Capacity     % Used    Path
    222.059 GB    4.9999 TB    4.44      /home/sia/storage
    219.211 GB    4.9999 TB    4.38      /home/sia/storage2
    222.093 GB    4.9999 TB    4.44      /home/sia/storage3

```


```
# netstat -tupan | grep LISTEN | grep siad
tcp        0      0 127.0.0.1:9980          0.0.0.0:*               LISTEN      849/siad        
tcp6       0      0 :::9981                 :::*                    LISTEN      849/siad        
tcp6       0      0 :::9982                 :::*                    LISTEN      849/siad 
```

```
siac wallet transactions | tail -n 40
      107977   3374dd7c5213ada759e3b6ff4dabcaed45e945f5d81aa4d99e4a60406bbec0d9          -1.34 SC             0 SF
      107978   26ebba961c0457d24592c569cc7b174890cd66986f622e8a31ea2e676d2a0228           0.00 SC             0 SF
      107978   bc357201a6fe56c04483956e5ef91a58b6eaaf9f88282f28650ae0400635552b          -1.34 SC             0 SF
      108031   f0a8f3b86a2fea5401e12580e441e3ae211b442c74b871a8a859459a028dab5a           0.00 SC             0 SF
      108031   a5dd30f5cdd837bcd19fad1a1e41ed87068326a8a7e9cb386f45fa5d3f23bfbd          -0.12 SC             0 SF
      108058   6515eed2b51f6aa13a258e32b1f011c72d543f50d99f8cfe14ca8640e844615b           0.00 SC             0 SF
      108058   d3aabc90be97f4b2800fa487b568a2834e7d94ab8242857089bfdbfec24f54ba          -1.40 SC             0 SF
      108063   23b03e70d4e8b861b718e902b2aea8df40d838a812b96886e3175650cfdf74a5           0.00 SC             0 SF
      108063   d7427c615fbc1d53b879825073a8a4ab51bd413263ca7c7b7a585576dda0bc22          -1.33 SC             0 SF
      108065   7d8da9d4d2cd9d2ba6947265c767f98bfd4eb6f17fa9c1fe8ddafd746e5ec1f4           0.00 SC             0 SF
      108065   83c94b98634819b0a8ba717487c0ee27e6933e2561284c2767379a238944bbb9         -16.63 SC             0 SF
      108097   114a787700080a804bfcc65eb54c208d4491c9cd410d76558759589d84bd703a           0.00 SC             0 SF
      108097   0597345347c63888fed509f9f8150ae161ad6b561fd0c5ff9616bb547f8d02e5          -1.17 SC             0 SF
      108108   ba95446e78394480cc5631169ba98ef26dc7c5284d482e9e42e0f5f4d8f9bb09           0.00 SC             0 SF
      108108   080c96d87139ebdccea76b17664bcfaad40de6e58aa3b4410b111c75a527139e          -2.97 SC             0 SF
      108120   91969dfe39f3b81bc0ea7c4b6b16d73f6f8aa9f1b513c7022fad1c4d0726806c           0.00 SC             0 SF
      108120   143787180ee9803cb5c2f7ae193fd50bb27efc249ef239e4d6e2f633e11cced4         -70.57 SC             0 SF
      108123   6feb64027bd8a951723daf8882289fe1f99e26e95ce83cc5cb613ec59d4e761b           0.00 SC             0 SF
      108123   3e8a8aafdb3dd1f1c84bb2356c22096982cd3ea0037016cd39059458c2d347f9          -1.04 SC             0 SF
      108124   7f71a6149700302566a1f844d4c3c6164953d7e1391ef8c37c2fdfc244e9c2a7           0.00 SC             0 SF
      108124   3b4d9dfde7ca651efce6f51817781b48032b4123426fc44c1446894189b66338          -1.30 SC             0 SF
      108125   08bc9042e579bec06b84ee373df59633601f4f52dcea695add195bb9e16c27da           0.00 SC             0 SF
      108125   1ff6b702d7b13f8ea8ef18a9fe87b15b768c396237742c5779698d0537c01571          -1.33 SC             0 SF
      108907   b1ab8bc512e84fcc4668780f3ba57b686c898ce6c378e13b3b29d3736bf604b2        -931.84 SC             0 SF
      108907   503fe6cb91f83069572f6ab5d0834d76f2c8756c4a7e3e78ab04897454b3127c        -710.94 SC             0 SF
      108907   f2b72965974380d64ab9f3ac5307ce55823a6c6e54d2d684a8810cd7c6abc561       -7825.62 SC             0 SF
      109274   a3e875319a7c270da981411c87551edf6a7ffb793b5f47a0b57cda38c51a9417           0.00 SC             0 SF
      109274   4b9800c5af6ae099df5927fc59fc0386b8b1ad9b451b28af566631cd9596ceb4          -1.34 SC             0 SF
      109274   34c3c2159bb14285010c0fd9c07944e08815edbc8f61f73a4ef616a54ae2a166           0.00 SC             0 SF
      109274   439e1f60305ec680197f33ac5ed38de1682bc9f8178f1e340724a89ad9505740          -1.34 SC             0 SF
      109275   780477d74b108916716c274a4c3963d7392b8a251791769ab54fdac657a084bc           0.00 SC             0 SF
      109275   cfa56a4df6b753882853edc6ce1dad0806a8a88a7bfda4563f2d8b4aa6bb01eb          -1.34 SC             0 SF
      109277   aec089fe7eaec94265893b26cada90e54e498d50246b65bf3ac740498ffdc95b           0.00 SC             0 SF
      109277   16245f3c829585657f8437b396d3f9c2144e76e10f5f053d5440ee2b20163013         -36.39 SC             0 SF
      109277   d682b441e6f1d01635e6f929d59bf4075988566d825be8bb4aa813954d888325           0.00 SC             0 SF
      109277   b003f5f63511500c4ffc4d5720ef125d176a1d203d2e6e2b7b68c96f113a2236        -162.16 SC             0 SF
      109279   cdd8c1a8defbebc5e3c2ec03d50646b50043b4a23feb917ac040649e4cef8f6b           0.00 SC             0 SF
      109279   d8ac0bf2b88202565361dfda202d58b280219e70999aad64d0281e2476f88534          -1.34 SC             0 SF
      109279   867a7f23c009f8d09c3cd932c01901bcfb2dbf35b4a4d76f56d676b9d82c5b64           0.00 SC             0 SF
      109279   420677ac3c9f76712e52ba48e3d2f7555d22cd102883ac102d99878fd5967b96          -1.33 SC             0 SF
```

**Comment by lukechampine** [comment]
This should be fixed in v1.3.0, please reopen if it's still reproducible.

**Comment by starius** [comment]
This happened here.
Sia Daemon v1.3.0

It looks like majority of goroutines are in `managedAddVirtualSector` function.

```
$ grep  '^goroutine ' sia/siad2.log | wc -l
22289
$ grep -w managedAddVirtualSector sia/siad2.log | wc -l
21506
$ grep '^goroutine '  sia/siad2.log | egrep -o '\[[a-z]+' | sort | uniq -c 
     48 [chan
  14711 [runnable
    188 [select
    446 [semacquire
     39 [sleep
   6805 [syscall
```
